packages/mobx-state-tree/src/core/node/object-node.ts

```diff
@@ -453,12 +453,17 @@ export class ObjectNode<C, S, T> extends BaseNode<C, S, T> {
         this.fireHook(Hook.beforeDetach)
         this.state = NodeLifeCycle.DETACHING
 
-        this.environment = this.root.environment // make backup of environment
-        this.identifierCache = this.root.identifierCache!.splitCache(this)
-        this.parent!.removeChild(this.subpath)
-        this.baseSetParent(null, "")
+        const newEnv = this.root.environment
+        const newIdCache = this.root.identifierCache!.splitCache(this)
 
-        this.state = NodeLifeCycle.FINALIZED
+        try {
```

Comment:
Just wonder, why did it become a try/catch?

Reply:
To recover properly in case the conversion to undefined upon detaching fails because it's not valid

Reply:
It will rethrow the error anyway, but at least it won't leave the node in an inconsistent state